### PR TITLE
feat: github entity provider

### PR DESCRIPTION
**Pull request by brentg-telus**
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [inline docs/integrations/github/discovery.md:84]: @Rugvip This is intended. Is there anyway to prevent this error?

Comment (pjungermann) [reply docs/integrations/github/discovery.md:84]: If you use code formatting, it will be ignored.

File: docs/integrations/github/discovery.md
Comment (pjungermann) [inline docs/integrations/github/discovery.md:22]:
```suggestion
package.
```

File: docs/integrations/github/discovery.md
Comment (pjungermann) [inline docs/integrations/github/discovery.md:79]:
```suggestion
This provider supports multiple organizations via unique provider IDs.
```

File: docs/integrations/github/discovery.md
Comment (pjungermann) [inline docs/integrations/github/discovery.md:104]: I think you refer to two different time settings. Above you mention refreshes every 100 seconds, but your example config which is also related to the scheduler frequency below is about every 35 minutes.

I guess you meant the internal processing loop above. This can be configured at the catalog.ts file, too, directly at the CatalogBuilder. The snippet below though is only about the discovery schedule.

File: docs/integrations/github/discovery.md
Comment (pjungermann) [inline docs/integrations/github/discovery.md:121]: Maybe, we should deprecate it already. What do you think @Rugvip? We did it with other processors, too. (Not Bitbucket as it is not fully replaced yet.)

**Review by pjungermann — COMMENTED** [review]
looks good! I left some minor comments about the documentation.

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [reply docs/integrations/github/discovery.md:84]: thanks for the info.

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [reply docs/integrations/github/discovery.md]: fixed

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [reply docs/integrations/github/discovery.md]: fixed

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [reply docs/integrations/github/discovery.md:107]: updated for clarity.

File: docs/integrations/github/discovery.md
Comment (pjungermann) [inline docs/integrations/github/discovery.md]: I meant to use **`catalogPath`** with code formatting like

```
**`catalogPath`**
```
instead of
```
**catalogPath**
```

I think for the documentation pages it is probably better if not the whole block is code formatted.

Comment (brentg-telus) [reply docs/integrations/github/discovery.md]: Fixed. Thanks for clarifying.

**Comment by blakeromano** [comment]
@brentg-telus @pjungermann I'd personally love to see an additional configuration option on this called something like `entitiesInYaml` for if I want to allow a yaml to have other things besides backstage entities. For example if I had this yaml 
```yaml
test:
  moreTest: true
backstage:
  entities:
  - BACKSTAGE ENTITY ONE
  - BACKSTAGE ENTITY TWO
```
That I could specify the path within the yaml to be `backstage.entities` and have it only look for backstage entities in the yaml under that section. That'd be SUPER helpful! If you think this is warranted in a separate PR/Issue feel free to let me know.

**Review by benjdlambert — COMMENTED** [review]
Some quick comments at first glance

File: plugins/catalog-backend-module-github/api-report.md
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/api-report.md:59]: That is how the BitbucketEntityProvider does it which is what was recommended as a base for this entity provider.

File: plugins/catalog-backend-module-github/api-report.md
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/api-report.md:55]: the logger is globally scoped to the class via this, no need to pass it to the function.

File: plugins/catalog-backend-module-github/api-report.md
Comment (pjungermann) [reply plugins/catalog-backend-module-github/api-report.md:59]: it is inlined at `BitbucketCloudEntityProvider`

```ts
  static fromConfig(
    config: Config,
    options: {
      logger: Logger;
      schedule: TaskRunner;
    },
  ): BitbucketCloudEntityProvider[] {
```

There is only a `BitbucketCloudEntityProviderConfig` type, but it is not exported.

See https://github.com/backstage/backstage/blob/master/plugins/catalog-backend-module-bitbucket-cloud/api-report.md

**Comment by pjungermann** [comment]
@blakeromano I think this is out of scope for this PR and more a general functionality disconnected from GitHub as source location for catalog files.
This provider will only discover catalog files and register a Location entity.

These will then be parsed at a later processing step.

And there is still the question whether this is desirable as a general feature.

You could create your own entity provider or provide a custom processor parser
```ts
CatalogBuilder.setEntityDataParser(parser: CatalogProcessorParser)
```
which defaults to `defaultEntityDataParser`.

With that you can create your own custom format. Even without YAML.

File: plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts
Comment (pjungermann) [inline plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts:136]: you can pass `class` here.

example
```ts
          const logger = this.logger.child({
            class: BitbucketCloudEntityProvider.prototype.constructor.name,
            taskId,
            taskInstanceId: uuid.v4(),
          });
```

File: plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts
Comment (pjungermann) [inline plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts:141]: you should pass the logger here. Currently, the new logger instance is unused.

File: plugins/catalog-backend-module-github/api-report.md
Comment (pjungermann) [reply plugins/catalog-backend-module-github/api-report.md:55]: actually, you create logger but don't use it at `createScheduleFn`. Left comments there.

File: plugins/catalog-backend-module-github/api-report.md
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/api-report.md:59]: My mistake, I was looking at the provider config file in the same dir. removed.

File: plugins/catalog-backend-module-github/api-report.md
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/api-report.md]: resolved

File: plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts]: resolved

File: docs/integrations/github/discovery.md
Comment (benjdlambert) [reply docs/integrations/github/discovery.md:120]: I'm up for deprecating this now, but maybe we wait until the next mainline release to deprecate it, as these docs are going to go live as soon as it's merged, but the Provider wont be available for some time yet in the mainline release at least.

File: plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts
Comment (brentg-telus) [reply plugins/catalog-backend-module-github/src/providers/GitHubEntityProvider.ts:119]: added

File: docs/integrations/github/discovery.md
Comment (brentg-telus) [reply docs/integrations/github/discovery.md:120]: Is there anything I need to do for the deprecation process?

Comment (benjdlambert) [reply docs/integrations/github/discovery.md:120]: Not right now at least. Let's create a seperate ticket to deprecate the Processor outside of this PR.

**Review by benjdlambert — APPROVED** [review]
OK - Let's go with this! Thanks for your hard work @brentg-telus and patience! ❤️

**Comment by brentg-telus** [comment]
> OK - Let's go with this! Thanks for your hard work @brentg-telus and patience! ❤️

Thanks for the patience on your side as well. Glad it's merged.

**Comment by regicsolutions** [comment]
@pjungermann do you know if there is any documentation on how we would be able to inject a custom processor parser to the GithubEntityProvider? i have done this in the past with the Bitbucket Provider given it has an optional parameter [options.parser](https://backstage.io/docs/integrations/bitbucketServer/discovery/#custom-location-processing) where a parser can be set.

**Comment by pjungermann** [comment]
@regicsolutions as there is no such option at that provider, you cannot use it. Seems like only `BitbucketServerEntityProvider` has this option. Neither `BitbucketCloudEntityProvider`, `GithubEntityProvider`, nor `GitlabDiscoveryEntityProvider` have it.

If you have a use case, I would recommend creating an issue for it and potentially, creating a PR to add this capability.

As this is only about the Location entity emitted, not sure how you use it for Bitbucket Server and how you intend to use it for GitHub.

**Pull request merged**
